Flask web app blog (#1664)
oktadeveloper/okta-blog: _source/_posts/2026-06-30-flask-oauth-web-app.md

Check out the complete source code on [GitHub](https://github.com/oktadev/flask-web-app-example) and get started without setting it up from scratch.

**Table of Contents**{: .hide }
* Table of Contents
{% include toc.md %}

## What you'll build

In this tutorial, you'll build a simple dashboard application and learn how to:

* Securely sign users in to view their profile information using Okta as the OpenID Connect (OIDC) identity provider.  
* Implement the Authorization Code flow with [Proof Key for Code Exchange (PKCE)]({% post_url 2025-07-28-express-oauth-pkce %}) to adhere to the latest OAuth 2.0 security best practices.  
* Add a feature for new users to self-register for an account directly from the sign-in page.  
* Enable authenticated users to interact with a protected API using OAuth 2.0 access tokens.

## Prerequisites

> * A recent version of [Python](https://www.python.org/) and [pip](https://pypi.org/project/pip/) installed  

Reviewer: Can you add, This blog is based on version 3.14.

1. Go to Applications \> Applications  
2. Click Create App Integration  
3. Select OIDC \- OpenID Connect as the sign-in method  
4. Select Web Application as the application type, then click Next  
5. Enter an app integration name  
6. Configure the redirect URIs:  
   * Sign-in redirect URIs: http://localhost:5000/authorization-code/callback  
   * Sign-out redirect URIs: http://localhost:5000  
7. In the Controlled Access section, select the appropriate access level  
8. Click Save

## Enable self-service user registration

User registration is vital for any application, and Okta makes the process quick and hassle-free. Setting this up involves two main steps:

1. Create a user profile enrollment policy \- This policy defines the attributes a user must provide when self-registering. See the [Okta documentation](https://help.okta.com/oie/en-us/content/topics/identity-engine/policies/create-profile-enrollment-policy.htm) for detailed instructions.  

> 2. Assign your application to the policy \- This action activates the self-service registration flow for your app. Follow the steps in the [product documentation](https://help.okta.com/oie/en-us/content/topics/identity-engine/policies/about-ssr.htm).

Reviewer: I think the steps on this page linked here are not clearly explained. Can we add the detailed steps here, like the ones you shared over chat? For the time being, users could be misled or misunderstood.

Author: sharing a detailed link to follow

Your Flask application now has secure user authentication and can use an access token to interact with a protected API.

## Learn more

If you'd like to learn more about the concepts covered in this tutorial, explore these official Okta resources:

* [Sign users in to your web app](https://developer.okta.com/docs/guides/sign-into-web-app-redirect/python/main/)  
* [Authorization Code flow with PKCE](https://developer.okta.com/docs/guides/implement-grant-type/authcodepkce/main/)  
* [Self-Service Registration](https://help.okta.com/oie/en-us/content/topics/identity-engine/policies/about-ssr.htm)  
* [Validate Access Tokens](https://developer.okta.com/docs/guides/validate-access-tokens/python/main/)  
* [Create an Authorization Server](https://developer.okta.com/docs/guides/customize-authz-server/main/)  
* [Secure a Node Express App with OAuth 2.0 and PKCE]({% post_url 2025-07-28-express-oauth-pkce %})

> Remember to connect with us on [LinkedIn](https://www.linkedin.com/company/oktadev), follow us on [X](https://twitter.com/oktadev), and subscribe to our [YouTube channel](https://www.youtube.com/c/OktaDev/) for more exciting content. We also want to hear from you about the topics you'd like to see and any questions you may have. Leave us a comment below!

Reviewer: As part of CTA, you can also add the Okta for developers LinkedIn link.

---
layout: blog_post
title: "How to Build a Secure Flask App with Okta and a Custom Resource Server"
author: akanksha-bhasin
by: advocate

> communities: [python,security]

Reviewer 2: ```suggestion
communities: [python]
```

The security label only applies to posts that are relevant to the security community. This post is not relevant to the entire security community.

description: "Learn to add secure authentication to a Flask app with Okta and call a protected API using scoped access tokens."

> tags: [python, flask, oauth, oidc, pkce, okta, authentication]

Reviewer 2: ```suggestion
tags: [python, flask, oauth, oidc, pkce, authentication]
```

okta should not be a tag. everything on this blog is okta related!

please also check that yuo aren't adding any new tags here. tags are expensive.

type: conversion
image: blog/flask-oauth-web-app/flask-oauth-web-app-social-image.jpg
github: https://github.com/oktadev/flask-web-app-example
---

> Python's simplicity and the flexibility of the Flask microframework make it a popular choice for quickly building web applications. While Flask provides the essentials to get you started, you'll need to implement a critical piece yourself: secure user authentication. After all, how do you securely sign users into your application?

Reviewer 2: ```suggestion
Python syntax and the flexibility of the Flask microframework make it a popular choice for quickly building web applications. While Flask provides the essentials to get you started, you'll need to implement a critical piece yourself: secure user authentication. After all, how do you securely sign users into your application?
```

Avoid insinuating something is simple or easy. it's insensitive

> This tutorial shows you how to solve that problem. You'll add a secure authentication layer to your Flask application using modern standards like OAuth 2.0 and OpenID Connect (OIDC). You'll also learn how to implement the secure Authorization Code flow with Proof Key for Code Exchange (PKCE) and build a complete sign-in and sign-out workflow. Second, and just as importantly, you'll learn how to use the resulting scoped access tokens to make protected calls to an API.

Reviewer 2: ```suggestion
This tutorial shows you how to solve that problem. You'll add a secure authentication layer to your Flask application using modern standards like OAuth 2.1 and OpenID Connect (OIDC). You'll also learn how to implement the secure Authorization Code flow with Proof Key for Code Exchange (PKCE) and build a complete sign-in and sign-out workflow. Second, and just as importantly, you'll learn how to use the resulting scoped access tokens to make protected calls to an API.
```

Check out the complete source code on [GitHub](https://github.com/oktadev/flask-web-app-example) and get started without setting it up from scratch.

**Table of Contents**{: .hide }
* Table of Contents
{% include toc.md %}

> ## What you'll build

Reviewer 2: Your section titles don't contribute to SEO. Use your keywords in your sections headers

In this tutorial, you'll build a simple dashboard application and learn how to:

> * Securely sign users in to view their profile information using Okta as the OpenID Connect (OIDC) identity provider.  

Reviewer 2: Do we need to define OIDC and PKCE acronyms twice?

> * Implement the Authorization Code flow with [Proof Key for Code Exchange (PKCE)]({% post_url 2025-07-28-express-oauth-pkce %}) to adhere to the latest OAuth 2.0 security best practices.  

Reviewer 2: Why are we referencing the post like this?

Also, i don't think your express app post is the appropriate link here. You should be referencing dev docs or oauth.com or some other actual reference about PKCE. If you want to plug your previous post using a different tech set, be up front about it (don't make someone navigate this link thinking it's a real reference though). I wouldn't do it in this section either way

* Add a feature for new users to self-register for an account directly from the sign-in page.  
* Enable authenticated users to interact with a protected API using OAuth 2.0 access tokens.

## Prerequisites

* [Python](https://www.python.org/) 3.14 or later and [pip](https://pypi.org/project/pip/) installed  
* An [Okta Integrator Free Plan](https://developer.okta.com/signup/)

## Create an app integration in the Okta Admin Console

Before you begin, you'll need an Okta Integrator Free Plan account. To get one, sign up for an [Integrator account](https://developer.okta.com/login). Once you have an account, sign in to your [Integrator account](https://developer.okta.com/login). Next, in the Admin Console:

> 1. Go to Applications \> Applications  

Reviewer 2: Please follow our standards. UI labels are bold

## Enable self-service user registration

User registration is vital for any application, and Okta makes the process quick and hassle-free. Setting this up involves two main steps:

1. Create a user profile policy \- This policy defines the attributes a user must provide when self-registering. See the [Okta documentation](https://help.okta.com/oie/en-us/content/topics/identity-engine/policies/create-profile-enrollment-policy.htm) for detailed instructions.  
2. Assign your application to the policy \- This step is required; without it, the self-service registration flow will not activate for your app. See the [product documentation](https://help.okta.com/oie/en-us/content/topics/identity-engine/policies/select-profile-enrollment-policy.htm) for more details.

## Add secure user login to a Flask application with Okta

Now that you've configured your app integration in Okta and noted the required Okta application details, it's time to create your Flask application.

### Create a basic Flask web application

First, you'll create a project folder and a Python virtual environment, then install the necessary dependencies.

> 1. Create a project folder named flask-web-app and a subfolder called venv.  

Reviewer 2: should these directory names be `inline code`?

2. Open a terminal window in the project folder and create a Python virtual environment

   ```bash
   python -m venv venv
   ```

3. Activate the virtual environment. Use the command that corresponds to your operating system.

   ```bash
   # On macOS/Linux
   source venv/bin/activate

   # On Windows
   venv\Scripts\activate.bat
   ```

4. With your environment active, install the required libraries:

   ```bash
   pip install Authlib==1.6.1 Flask[async]==3.1.1 python-dotenv==1.1.1 requests==2.32.4 okta-jwt-verifier==0.3.0
   ```

### Configure Flask environment variables

To keep sensitive credentials secure, you'll store them in a .env file and load them into the application's configuration. This practice prevents hard-coding secrets directly into your source code.

1. Create a file named .env in the project root and add your Okta application's configuration details. Replace the placeholders with your actual values.

   ```bash
   OKTA_CLIENT_ID={yourClientId}
   OKTA_CLIENT_SECRET={yourClientSecret}
   OKTA_BASE_URL=https://{yourOktaDomain}
   ```

> 2. Create a config.py file in the project root. This file reads the values from .env and constructs the OIDC metadata URL, which the application uses to fetch endpoints from Okta.

Reviewer 2: really feel like you should use inline code for filenames and such. it makes it easier for a reader to follow

from authlib.integrations.flask_client import OAuth
from dotenv import load_dotenv
import os
import requests

load_dotenv()

app = Flask(__name__)
app.secret_key = os.urandom(24)

app.config.from_object('config')

@app.route("/")
def index():
    return render_template('index.html')

if __name__ == "__main__":
    app.run(debug=True)
```

#### Initialize the Okta OIDC client

To set up the connection to Okta, add the following code to the bottom of your app.py file. This code initializes an OAuth object and registers the Okta application as a remote provider.

```python
# Create a registry with an OAuth object
oauth = OAuth(app)

# Register a remote application
oauth.register(
    name='okta',
    client_id=app.config['OKTA_CLIENT_ID'],
    client_secret=app.config['OKTA_CLIENT_SECRET'],
    server_metadata_url=app.config['OKTA_METADATA_URL'],
    client_kwargs={
        'scope': 'openid profile email offline_access',
        'code_challenge_method': 'S256'
    }
)
```

The first parameter in the register() method (in this case, *Okta*) is the name of the remote application. You'll later access the remote application with oauth.okta to handle all the OIDC-related functions.

> The client\_kwargs dictionary passes extra parameters to the authorization request. Because the code\_challenge\_method is present, the Authlib library automatically uses the [Proof Key for Code Exchange (PKCE)]({% post_url 2025-07-28-express-oauth-pkce %}#why-use-pkce-in-oauth-20) flow, which enhances security.

Reviewer 2: once again, please follow our styles for referencing other posts.

Also your express post is not the definitive pkce guide. refer to an actual reference material.

#### Build Flask routes for OIDC authentication

The sign-in process begins when the user clicks the Login button, which redirects the browser to the Okta-hosted Sign-In page.

After the user signs in, Okta redirects the browser to the configured sign-in redirect URI, sending user profile information to the application. Similarly, after a user signs out, Okta redirects the browser to the sign-out redirect URI.

Let's start by creating a simple HTML page that allows the user to log in. This page will display a sign-in button for signed-out users and a sign-out button for logged-in users, along with profile information.

> 1. Add the /login route to your app.py file. This endpoint serves as the entry point for the authentication flow. It constructs the authorization redirect URI and sends the user's browser to the Okta-hosted Sign-In page to begin the OIDC process.

Reviewer 2: use inline code
